### DISCUSSION TITLE
[resolved] 0.36.0/3.5.0 Zookeeper can't find other zookeepers

### DISCUSSION THREAD

**Discussion by maartengo** [strimzi/strimzi-kafka-operator#8890] [General]
On a fresh deployment of the strimzi-kafka-operator and kafka the zookeerper pods log the following. 

> ERROR StaticHostProvider:148 - Unable to resolve address: dev-dev2-zookeeper-1.dev-dev2-zookeeper-nodes.kafka.svc/<unresolved>:2181

This seems to block the rest of the kafka deployment, only zookeeper pods were created.

We've deployed to the `kafka` namespace, and the service has the name `dev-dev2-zookeeper-nodes`.

Our deployment steps;
0. Make sure no existing kafka configuration is present, delete `kafka` namespace
1. Deploy version `0.36.0` of the operator chart to `kafka` namespace
2. Wait for deployment to complete
3. Deploy version `3.5.0` of kafka
4. Wait for deployment to complete
5. See that only zookeeper pods are created
6. Open zookeeper logs
7. See something similar to [zookeeper-logs.txt](https://github.com/strimzi/strimzi-kafka-operator/files/12146066/zookeeper-logs.txt)

Operator Helm chart; 0.36.0
Kubernetes; 1.26

---

Do we need to supply different values/configuration for the deployment of `3.5.0`? What could be going wrong here?

**Comment by scholzj** [comment]
This looks like it might be related to the new ZooKeeper version used by KAfka 3.5 which seems to be a lot more sensitive to the quality of your Kubernetes DNS.

CC @ppatierno

**Reply by ppatierno** [reply]
Yeah it seems to be the issue. It could take more time to come up, sometimes. Were the Kafka brokers created eventually? 
Can you also try if killing one ZK pod would help?

CC @fvaleri @showuon

**Reply by maartengo** [reply]
Only the ZK pods are currently created, killing one of the pods also doesn't change the state.

**Reply by ppatierno** [reply]
@maartengo can you give the following a try ...

Create a new Kafka cluster, by specifying the following image for the ZooKeeper `quay.io/ppatierno/kafka:0.36.0-kafka-3.5.0-dns-check`.
Without changing anything in the operator deployment, you can do that in the `zookeeper.image` field in your `Kafka` custom resource.
So something like this:

```yaml
# ....
zookeeper:
    replicas: 3
    image: quay.io/ppatierno/kafka:0.36.0-kafka-3.5.0-dns-check
    storage:
# .....
```


Please, let me know if creating a Kafka cluster from scratch by using this image helps somehow.

**Reply by showuon** [reply]
@maartengo , thanks for reporting the issue!

1. It would be great if you can collect all the zookeeper pod's log for investigation next time when zookeeper stuck again. It could be done by command like this:
```
kubectl logs dev-dev2-zookeeper-0 -n kafka > /tmp/zk0.log
```

2. From your log, I can see the `zookeeper-0` pod is created at 12:16:26, and the log ends at 12:17:06, which only stay for 40 seconds for it to be stable. The issue sometimes can be self-healing after 1 ~ 2 min. I'd suggest you wait for 1 ~ 2 mins before collecting logs, and see if it can be self-healing. 

Thank you.

**Reply by maartengo** [reply]
With that image I no longer see the `Unable to resolve address` error, but the ZK pods still log exceptions and no other pods are created. The following warning is shown before a ZK Down exception `Closing connection to /10.120.16.74:57070`. That IP is one of the strimzi operators.
[zookeeper-custom-image.txt](https://github.com/strimzi/strimzi-kafka-operator/files/12157971/zookeeper-custom-image.txt)

**Reply by maartengo** [reply]
I only showed the first few logs since after that it keeps repeating, but I'll try to post more logs in the future as well. The logs after ~9 minutes for both replicas: 
[zookeeper-custom-image-rep1.txt](https://github.com/strimzi/strimzi-kafka-operator/files/12158038/zookeeper-custom-image-rep1.txt)
[zookeeper-custom-image-rep2.txt](https://github.com/strimzi/strimzi-kafka-operator/files/12158039/zookeeper-custom-image-rep2.txt)

**Reply by ppatierno** [reply]
@maartengo why do you have only 2 ZooKeeper? This is what I see from the server configuration in the logs. This is really unusual, 3 or 5 are usually used.

**Reply by ppatierno** [reply]
@maartengo can you also share the `Kafka` custom resource, please?

**Reply by ppatierno** [reply]
@maartengo I have also updated the `quay.io/ppatierno/kafka:0.36.0-kafka-3.5.0-dns-check` image to print more log I could need for investigation. Can you give it another try please? Be sure to delete the older image from your environment first.

**Reply by maartengo** [reply]
We use 2 replicas because this is a small K8S cluster and we mostly just want to test that everything can deploy OK. I can run a test with more replicas, though I doubt that causes the problem. 

[kafka-manifest.txt](https://github.com/strimzi/strimzi-kafka-operator/files/12158515/kafka-manifest.txt)

**Reply by maartengo** [reply]
@ppatierno could you publish the image with a new version? I have no easy way of removing images from the environment

**Reply by ppatierno** [reply]
@maartengo just pushed `quay.io/ppatierno/kafka:0.36.0-kafka-3.5.0-dns-check-v2`
Btw I tried with your 2 ZK replicas configuration and I have the problem. With 2 replicas it's possible not reaching the quorum and having split brain, or? It's possible but not sure. With this new ZK version which is sensitive to DNS resolution, it could happen more. With 3 ZK it works just fine for me.

**Reply by fvaleri** [reply]
@maartengo can you try with just 1 or 3 ZK replicas? I think with only two nodes it may take a long time to elect a stable leader, because every node will vote for itself.

**Reply by fvaleri** [reply]
Actually, I see 60 leader elections in the last logs you shared, each one followed by connection errors and then a timeout while waiting for epoch from quorum, which triggers a new election round. This goes on and on, so it could still be caused by the slow DNS resolution problem.

**Reply by maartengo** [reply]
Just tried again with 2 replicas and the new image version:
[zookeeper-custom-v2-rep1.txt](https://github.com/strimzi/strimzi-kafka-operator/files/12159253/zookeeper-custom-v2-rep1.txt)
[zookeeper-custom-v2-rep2.txt](https://github.com/strimzi/strimzi-kafka-operator/files/12159254/zookeeper-custom-v2-rep2.txt)

When running with 3 replicas the problem does not occur, all kafka resources are created as expected. For me this seems to indicate that the problem isn't dns, since then it shouldn't work no matter the replica count.
[zookeeper-custom-v2-3reps-rep1.txt](https://github.com/strimzi/strimzi-kafka-operator/files/12159319/zookeeper-custom-v2-3reps-rep1.txt)

I understand running with 1/3/5 replicas is the norm, but I can imagine that this issue should still be addressed in case of a disruption where one of the ZK instances does not properly respond, right? Or would it then not occur?

**Reply by ppatierno** [reply]
@maartengo can you give it another try with 2 ZKs but using this image please? `quay.io/ppatierno/kafka:0.36.0-kafka-3.5.0-all-binding-gt-2`

**Reply by maartengo** [reply]
Sure, here are the logs for ZK with 2 replicas:
[zookeeper-custom-v3-rep2.txt](https://github.com/strimzi/strimzi-kafka-operator/files/12160297/zookeeper-custom-v3-rep2.txt)
[zookeeper-custom-v3-rep1.txt](https://github.com/strimzi/strimzi-kafka-operator/files/12160299/zookeeper-custom-v3-rep1.txt)

This only spawned the ZK pods and no others.

**Reply by ppatierno** [reply]
@maartengo I tested this image on different envs and it works with 2 ZKs. In your logs I don't see ZK errors anymore so they should have reached the quorum this time, so it would be interesting to get the cluster operator log at the same date/time to get why the Kafka clusters are not showing up. I also assume you waited enough.

**Reply by fvaleri** [reply]
@maartengo could you share the full env setup using [report.sh](https://raw.githubusercontent.com/strimzi/strimzi-kafka-operator/main/tools/report.sh) and attach the result?

**Reply by maartengo** [reply]
[reports.zip](https://github.com/strimzi/strimzi-kafka-operator/files/12161003/reports.zip)

**Reply by ppatierno** [reply]
@maartengo I guess your cluster operator is installed in a different namespace from where Kafka is running so the reports didn't get its log. Can you provide the cluster operator log please? It's the most important one we need :-)

**Reply by maartengo** [reply]
My apologies, I thought those would be included since the operator _is_ running in the same namespace. The logs can be found below; 
[strimzi-cluster-operator-2.txt](https://github.com/strimzi/strimzi-kafka-operator/files/12161206/strimzi-cluster-operator-2.txt)
[strimzi-cluster-operator-1.txt](https://github.com/strimzi/strimzi-kafka-operator/files/12161207/strimzi-cluster-operator-1.txt)

**Reply by ppatierno** [reply]
@maartengo ZK is working fine and CO is able to connect if you check the log, now you have a different problem as you can see there ...

```shell
2023-07-25 10:11:36 ERROR KafkaListenersReconciler:765 - Reconciliation #302(watch) Kafka(kafka/ismcpsdev-dev2): Failed to process Secrets with custom certificates: [Secret custom-external-broker-cert with custom TLS certificate does not exist.]
2023-07-25 10:11:36 ERROR AbstractOperator:276 - Reconciliation #302(watch) Kafka(kafka/ismcpsdev-dev2): createOrUpdate failed
io.strimzi.operator.cluster.model.InvalidResourceException: Failed to process Secrets with custom certificates: [Secret custom-external-broker-cert with custom TLS certificate does not exist.]
	at io.strimzi.operator.cluster.operator.assembly.KafkaListenersReconciler.lambda$customListenerCertificates$44(KafkaListenersReconciler.java:766) ~[io.strimzi.cluster-operator-0.36.0.jar:0.36.0]
	at io.vertx.core.impl.future.Composition.onSuccess(Composition.java:38) ~[io.vertx.vertx-core-4.4.4.jar:4.4.4]
```

a Secret missing while it's configured in your Kafka custom resource as here:

```yaml
brokerCertChainAndKey:
          certificate: tls.crt
          key: tls.key
          secretName: custom-external-broker-cert
```

**Reply by fvaleri** [reply]
Yeah, was about to post the same.

```sh
  - lastTransitionTime: "2023-07-25T11:54:36.184587765Z"
    message: 'Failed to process Secrets with custom certificates: [Secret custom-external-broker-cert
      with custom TLS certificate does not exist.]'
    reason: InvalidResourceException
    status: "True"
    type: NotReady
```

**Reply by maartengo** [reply]
Ah right, I deleted the environment too many times and now we can't request a new certificate 🙈
I'll deploy it on another environment and will report back with the results

**Reply by maartengo** [reply]
Just tested on a new environment; the `quay.io/ppatierno/kafka:0.36.0-kafka-3.5.0-all-binding-gt-2` image does indeed work when running 2 replicas of ZK

**Reply by ppatierno** [reply]
@maartengo thanks for the feedback. The fix should be included in the next 0.37.0 which won't be too soon (we just released 0.36.0) so my advice for you is to run at least 3 ZK because this private image I provided you for testing will soon go away as well and it's anyway not "official".

**Reply by maartengo** [reply]
Is it possible to include the fix in a patch release? It would be a shame if we'd have to wait for the next release just so we can run a certain amount of replicas. I can also image that this issue occurs for more users of the operator, who would all have to find out what exactly is going wrong before knowing to change their ZK replica count.

**Reply by ppatierno** [reply]
> Is it possible to include the fix in a patch release?

@scholzj we are planning a 0.36.1, do you think we can include this fix?

>  I can also image that this issue occurs for more users of the operator

Actually ZK with 2 nodes is a really bad a not so used configuration. We totally discourage it and you can also see a warning in the Kafka custom resource status about it :-). I am not sure how many users really use this.

**Comment by showuon** [comment]
> I only showed the first few logs since after that it keeps repeating, but I'll try to post more logs in the future as well. The logs after ~9 minutes for both replicas: [zookeeper-custom-image-rep1.txt](https://github.com/strimzi/strimzi-kafka-operator/files/12158038/zookeeper-custom-image-rep1.txt) [zookeeper-custom-image-rep2.txt](https://github.com/strimzi/strimzi-kafka-operator/files/12158039/zookeeper-custom-image-rep2.txt)

For the record, I've opened a [JIRA issue](https://issues.apache.org/jira/browse/ZOOKEEPER-4724) in Zookeeper's project. I believe there is a bug there, but need zookeeper's expert to identify it.
